x/emissions/module/rewards_v2.go

```diff
@@ -0,0 +1,430 @@
+	if len(stakes) != len(scores) {
+		return nil, types.ErrInvalidSliceLength
+	}
+
+	// Calculate (stakes * scores)^preward and sum of all fractions
+	var totalFraction float64
+	fractions := make([]float64, len(stakes))
+	for i, stake := range stakes {
+		fractions[i] = math.Pow(stake*scores[i], preward)
+		totalFraction += fractions[i]
+	}
+
+	// Normalize fractions
+	for i := range fractions {
+		fractions[i] /= totalFraction
+	}
+
+	return fractions, nil
+}
+
+// GetfUniqueAgg calculates the unique value or impact of each forecaster.
+// ƒ^+
+func GetfUniqueAgg(numForecasters float64) float64 {
+	return 1.0 / math.Pow(2.0, (numForecasters-1.0))
+}
+
+// GetFinalWorkerScoreForecastTask calculates the worker score in forecast task.
+// T_ik
+func GetFinalWorkerScoreForecastTask(scoreOneIn, scoreOneOut, fUniqueAgg float64) float64 {
+	return fUniqueAgg*scoreOneIn + (1-fUniqueAgg)*scoreOneOut
+}
+
+// GetWorkerScore calculates the worker score based on the losses and lossesCut.
+// Consider the staked weighted inference loss and one-out loss to calculate the worker score.
+// T_ij / T^-_ik / T^+_ik
+func GetWorkerScore(losses, lossesOneOut float64) float64 {
+	deltaLogLoss := math.Log10(lossesOneOut) - math.Log10(losses)
+	return deltaLogLoss
+}
+
+// GetStakeWeightedLoss calculates the stake-weighted average loss.
+// Consider the losses and the stake of each reputer to calculate the stake-weighted loss.
+// The stake weighted loss is used to calculate the network-wide losses.
+// L_i / L_ij / L_ik / L^-_i / L^-_il / L^+_ik
+func GetStakeWeightedLoss(reputersStakes, reputersReportedLosses []float64) (float64, error) {
+	if len(reputersStakes) != len(reputersReportedLosses) {
+		return 0, types.ErrInvalidSliceLength
+	}
+
+	totalStake := 0.0
+	for _, stake := range reputersStakes {
+		totalStake += stake
+	}
+
+	var stakeWeightedLoss float64 = 0
+	for i, loss := range reputersReportedLosses {
+		weightedLoss := (reputersStakes[i] * math.Log10(loss)) / totalStake
+		stakeWeightedLoss += weightedLoss
+	}
+
+	return stakeWeightedLoss, nil
+}
+
+// GetStakeWeightedLossMatrix calculates the stake-weighted geometric mean of the losses to generate the consensus vector.
+// L_i - consensus loss vector
+func GetStakeWeightedLossMatrix(reputersAdjustedStakes []float64, reputersReportedLosses [][]float64) ([]float64, error) {
+	if len(reputersAdjustedStakes) == 0 || len(reputersReportedLosses) == 0 {
+		return nil, types.ErrInvalidSliceLength
+	}
+
+	// Calculate total stake for normalization
+	totalStake := 0.0
+	for _, stake := range reputersAdjustedStakes {
+		totalStake += stake
+	}
+
+	// Ensure every loss array is non-empty and calculate geometric mean
+	stakeWeightedLoss := make([]float64, len(reputersReportedLosses[0]))
+	for j := 0; j < len(reputersReportedLosses[0]); j++ {
+		logSum := 0.0
+		for i, losses := range reputersReportedLosses {
+			logSum += (math.Log10(losses[j]) * reputersAdjustedStakes[i]) / totalStake
```

Comment:
a super easy way to save on compute that applies here + anywherhe else where there is a normalization (plenty of places) is to calculate the sum as you're looping through this then divide the weighted sum in the end. So 2 for-loops turns into 1, a division at every iteration turns into just 1, and you just increment 2 variables as you go along: the weighted sum (numerator) and the normalization constant (denominator)

Reply:
not sure if it's applicable here, since I'm calculated that stake weighted loss for each set of reported losses.